torchtune/models/gemma2/_attention.py

```diff
@@ -0,0 +1,320 @@
+
+        # Normalize q
+        if self.q_norm is not None:
+            q = self.q_norm(q)
+
+        if y is None:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "Must provide y input or use kv_cache to enable streaming decoding"
+                )
+            k = self.kv_cache.k_cache
+            v = self.kv_cache.v_cache
+        else:
+            # Update k and v shape, positional embeddings, and normalization
+
+            # k has shape [b, s_y, num_kv_heads * head_dim]
+            # v has shape [b, s_y, num_kv_heads * head_dim]
+            k = self.k_proj(y)
+            v = self.v_proj(y)
+
+            # Apply positional embeddings
+            # k: [b, s_y, n_kv, h_d]
+            k = k.view(b, s_y, -1, self.head_dim)
+            if self.pos_embeddings is not None:
+                k = self.pos_embeddings(k, input_pos=input_pos)
+
+            # View + expand + reshape bring num_kv_heads to num_heads for k and v
+            # to match q.
+
+            # k: [b, s_y, n_kv, 1, h_d]
+            # v: [b, s_y, n_kv, 1, h_d]
+            k = k.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
+            v = v.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
+
+            # If needed, expand the key and value tensors to have the same shape
+            # as the query tensor by copying values across the relevant dim
+            if self.num_heads != self.num_kv_heads:
+                k = k.expand(b, s_y, self.num_kv_heads, q_per_kv, self.head_dim)
+                v = v.expand(b, s_y, self.num_kv_heads, q_per_kv, self.head_dim)
+
+            # [b, s, n_h, h_d]
+            k = k.reshape(b, s_y, -1, self.head_dim)
+            v = v.reshape(b, s_y, -1, self.head_dim)
+
+            # [b, n_h, s, h_d]
+            k = k.transpose(1, 2)
+            v = v.transpose(1, 2)
+
+            # Normalize k
+            if self.k_norm is not None:
+                k = self.k_norm(k)
+
+            # Update key-value cache
+            if self.kv_cache is not None and self.cache_enabled:
+                k, v = self.kv_cache.update(k, v)
+
+        q.mul_(self.scaling)
+        output = torch.matmul(q, k.transpose(2, 3))
+
+        # if mask is None: default to causal mask
+        if mask is None:
+            mask = torch.tril(
+                torch.ones(
+                    size=(s_x, s_x),
+                    dtype=torch.bool,
+                ).to(x.device)
+            )
+
+        if self.sliding_window_size is not None:
+            all_ones = torch.ones_like(mask)
+
+            sliding_mask = torch.triu(
+                all_ones, -1 * self.sliding_window_size + 1
+            ) * torch.tril(all_ones, self.sliding_window_size - 1)
+            mask = torch.where(sliding_mask == 1, mask, -2.3819763e38)
+
+        if self.softcapping is not None:
+            output = output / self.softcapping
+            output = torch.tanh(output)
+            output = output * self.softcapping
```

Comment:
Can we add code comments explaining sliding window and the softcapping? (Also one for the magic value in the `torch.where` line wouldn't hurt)

Reply:
About this part of the code, I actually followed blindly the official pytoch implementation from Google [here](
    https://github.com/google/gemma_pytorch/blob/80881c2e6e797ef1913a4a705d4b40394791cc58/gemma/model.py#L213)

I am not sure why they used this magic number instead of `-torch.inf` ...

About the `sliding_mask` I am now worried that something is wrong here because of the way I defined the causal mask...

```
s_x = 10
sliding_window_size = 5
mask = torch.tril(
                torch.ones(
                    size=(s_x, s_x),
                    dtype=torch.bool,
                )
            )
print(mask)

all_ones = torch.ones_like(mask)
sliding_mask = torch.triu(
                all_ones, -1 * sliding_window_size + 1
            ) * torch.tril(all_ones, sliding_window_size - 1)
mask = torch.where(sliding_mask == 1, mask, -2.3819763e38)
print(mask)
```
The final mask here  does not seem to be causal anymore, and sliding future tokens are now accessible somehow...

Something like the following would seem better to me but is there a difference in the way masks are defined in gemma2 official code and torchtune?

```
s_x = 10
sliding_window_size = 5
mask = torch.tril(
                torch.ones(
                    size=(s_x, s_x),
                    dtype=torch.bool,
                )
            )

mask = torch.where(mask==0, -torch.inf, 1)
print(mask)

all_ones = torch.ones_like(mask)
sliding_mask = torch.triu(
                all_ones, -1 * sliding_window_size + 1
            ) * torch.tril(all_ones, sliding_window_size - 1)
mask = torch.where(sliding_mask == 1, mask, -torch.inf)
print(mask)
```

This seems concerning... what do you think ?

Reply:
@Optimox yeah good catch, that first mask definitely does not look right. I'm not that familiar with the official implementation, but looks like they are treating it as an additive mask [here](https://github.com/google/gemma_pytorch/blob/80881c2e6e797ef1913a4a705d4b40394791cc58/gemma/model.py#L321). In that case there should definitely not be anything that's not -torch.inf (or a very large negative number) above the diagonal.

This seems like a bug to me, maybe you can open an issue on the gemma_pytorch repo to confirm? Your second implementation looks correct to me.

Reply:
Sorry to add another kitchen to the cook here so late. I thought I'd share some of my findings from the last time I worked on masking + SDPA. I'd reccomend checking this PR out https://github.com/pytorch/pytorch/pull/133882 and the linked issues. 

TLDR; this is the correct approach to use the attention mask like a "bias" by adding a very large negative to the q/k tensors. However, using `-inf` as this negative number has been shown to produce NaN gradients for some rare corner cases (e.g. when an entire row is masked out). In Transformers, the approach is to use something like `torch.finfo(dtype).min` [1] (which is maybe where the original magic number is coming from?)
```python
import torch
x = torch.Tensor([[[float("-inf"), float("-inf"), float("-inf"), 10]]])
softmax = torch.nn.Softmax(dim=-1)
softmax(x)
# tensor([[[nan, nan, nan]]])
dtype = torch.bfloat16
min_value = torch.finfo(dtype).min
# -3.3895313892515355e+38 - on MPS, this will vary depending on the hardware you're using
x = torch.Tensor([[[min_value, min_value, min_value]]])
softmax = torch.nn.Softmax(dim=-1)
softmax(x)
# tensor([[[0.3340, 0.3340, 0.3340]]]) 
```



Aside, as of torch 2.5 this is handled [internally](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L669) slightly differently. `-inf` is used in the mask, softmax is performed, but then any rows in the original tensor which have entirely masked out rows are explicitly set to zero.



[1] Transformers follows this approach for their [Gemma2 implementation](https://github.com/huggingface/transformers/blob/33868a057c02f0368ba63bd1edb746be38fe3d90/src/transformers/models/gemma2/modeling_gemma2.py#L545). However, this apparently still causes issues for some dtypes so it's also been suggested to use `torch.finfo(dtype).min / 2` - see https://github.com/huggingface/transformers/issues/32390, or to just attend to all tokens in a row containing only padding tokens equally (https://github.com/huggingface/transformers/commit/e22d913e8b9717d6a51b2e67e08c2f3bd9577a48), but I'm not 100% about how this works.

Reply:
Wait so I am a bit confused here (this is also in reference to the issue opened by @Optimox on the gemma_pytorch repo). There are two separate questions here, right?

(1) Should values above the diagonal be unmasked? 
(2) What is the right masked value for an additive mask?

I think both the gemma_pytorch discussion and @SalmanMohammadi's comments address (2), and that I am not so worried about. But I think (1) is more fundamental, and I can directly copy-paste the [snippet from gemma_pytorch](https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L312-L316) to get this:

```
import torch

s_x = 5
sliding_window_size = 3
mask = torch.tril(
                torch.ones(
                    size=(s_x, s_x),
                    dtype=torch.bool,
                )
            )

all_ones = torch.ones_like(mask)
sliding_mask = torch.triu(
    all_ones, -1 * sliding_window_size + 1
) * torch.tril(all_ones, sliding_window_size - 1)
mask = torch.where(sliding_mask == 1, mask, -2.3819763e38)
print(mask)
...

tensor([[ 1.0000e+00,  0.0000e+00,  0.0000e+00, -2.3820e+38, -2.3820e+38],
        [ 1.0000e+00,  1.0000e+00,  0.0000e+00,  0.0000e+00, -2.3820e+38],
        [ 1.0000e+00,  1.0000e+00,  1.0000e+00,  0.0000e+00,  0.0000e+00],
        [-2.3820e+38,  1.0000e+00,  1.0000e+00,  1.0000e+00,  0.0000e+00],
        [-2.3820e+38, -2.3820e+38,  1.0000e+00,  1.0000e+00,  1.0000e+00]])
```

This demonstrates that there are values above the diagonal that are unmasked, no? (I will also reopen the issue on there just to confirm I am not missing something here)

Reply:
What I understood is that there are two ways of defining masks (in bias mode):
- 0 for reachable, -inf (or very large negative number) for unreachable -> this is the expected input mask for gemma_pytorch implementation
- bolean mask (True for reachable, False otherwise) + [torch.nn.functional.scaled_dot_product_attention](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) which internally switches from boolean to the first implementation `if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))`

So here, since all previous models from torchtune used torch sdpa there was a mismatch between both implementation, that is why I added this conversion in the latest [changes](https://github.com/pytorch/torchtune/pull/1835/commits/7adb55a4da1b07bdc1dc56aff79d6446a96aee9e). This conversion should also work for block mask defined as boolean mask currently in torchtune code.

Reply:
Thanks, I also caught up with @SalmanMohammadi offline about this. It seems like the initial definition of `mask` in the code snippet I shared in my last comment did not match what they do in gemma_pytorch, so that was a misunderstanding on my part. I think your approach makes sense, will look at the code more closely to confirm the BlockMask case (though I guess if we're not supporting packed or flex yet it doesn't matter?)